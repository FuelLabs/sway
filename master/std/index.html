<!DOCTYPE html><html><head><meta charset="utf-8"><meta name="viewport" content="width=device-width, initial-scale=1.0"><meta name="generator" content="swaydoc"><meta name="description" content="API documentation for the Sway `std` module in `std`."><meta name="keywords" content="sway, swaylang, sway-lang, std"><link rel="icon" href="../static.files/sway-logo.svg"><title>std in std - Sway</title><link rel="stylesheet" type="text/css" href="../static.files/normalize.css"><link rel="stylesheet" type="text/css" href="../static.files/swaydoc.css" id="mainThemeStyle"><link rel="stylesheet" type="text/css" href="../static.files/ayu.css"><link rel="stylesheet" href="../static.files/ayu.min.css"></head><body class="swaydoc mod"><nav class="sidebar"><a class="sidebar-logo" href="#"><div class="logo-container"><img class="sway-logo" src="../static.files/sway-logo.svg" alt="logo"></div></a><h2 class="location">Library std</h2><div class="sidebar-elems"><a id="all-types" href="all.html"><p>See all std's items</p></a><section><div class="block"><ul><li><a href="#modules">Modules</a></li><li><a href="#primitives">Primitives</a></li></ul></div></section></div></nav><main><div class="width-limiter"><script src="../search.js" type="text/javascript"></script><script>function onSearchFormSubmit(event){event.preventDefault();const searchQuery=document.getElementById("search-input").value;const url=new URL(window.location.href);if(searchQuery){url.searchParams.set('search',searchQuery)}else{url.searchParams.delete('search')}history.pushState({search:searchQuery},"",url);window.dispatchEvent(new HashChangeEvent("hashchange"))}document.addEventListener('DOMContentLoaded',()=>{const searchbar=document.getElementById("search-input");const searchForm=document.getElementById("search-form");searchbar.addEventListener("keyup",function(event){onSearchFormSubmit(event)});searchbar.addEventListener("search",function(event){onSearchFormSubmit(event)});function onQueryParamsChange(){const searchParams=new URLSearchParams(window.location.search);const query=searchParams.get("search");const searchSection=document.getElementById('search');const mainSection=document.getElementById('main-content');const searchInput=document.getElementById('search-input');if(query){searchInput.value=query;const results=Object.values(SEARCH_INDEX).flat().filter(item=>{const lowerQuery=query.toLowerCase();return item.name.toLowerCase().includes(lowerQuery)});const header=`<h1>Results for ${query}</h1>`;if(results.length>0){const resultList=results.map(item=>{const formattedName=`<span class="type ${item.type_name}">${item.name}</span>`;const name=item.type_name==="module"?[...item.module_info.slice(0,-1),formattedName].join("::"):[...item.module_info,formattedName].join("::");const path=["..",...item.module_info,item.html_filename].join("/");const left=`<td><span>${name}</span></td>`;const right=`<td><p>${item.preview}</p></td>`;return`<tr onclick="window.location='${path}';">${left}${right}</tr>`}).join('');searchSection.innerHTML=`${header}<table>${resultList}</table>`}else{searchSection.innerHTML=`${header}<p>No results found.</p>`}searchSection.setAttribute("class","search-results");mainSection.setAttribute("class","content hidden")}else{searchSection.setAttribute("class","search-results hidden");mainSection.setAttribute("class","content")}}window.addEventListener('hashchange',onQueryParamsChange);onQueryParamsChange()})</script><nav class="sub"><form id="search-form" class="search-form" onsubmit="onSearchFormSubmit(event)"><div class="search-container"><input id="search-input" class="search-input" name="search" autocomplete="off" spellcheck="false" placeholder="Search the docs..." type="search"></div></form></nav><section id="main-content" class="content"><div class="main-heading"><h1 class="fqn"><span class="in-band">Library <a class="mod" href="#">std</a></span></h1></div><h2 id="modules">Modules</h2><div class="item-table"><div class="item-row"><div class="item-left Module-item"><a class="mod" href="address/index.html">address</a></div><div class="item-right docblock-short"><p>The <code>Address</code> type used for interacting with addresses on the fuel network.</p>
</div></div><div class="item-row"><div class="item-left Module-item"><a class="mod" href="alloc/index.html">alloc</a></div><div class="item-right docblock-short"><p>A library for allocating memory inspired by <a href="https://doc.rust-lang.org/std/alloc/index.html">Rustâ€™s std::alloc</a>.</p></div></div><div class="item-row"><div class="item-left Module-item"><a class="mod" href="assert/index.html">assert</a></div><div class="item-right docblock-short"><p>Functions to assert a given condition.</p>
</div></div><div class="item-row"><div class="item-left Module-item"><a class="mod" href="asset/index.html">asset</a></div><div class="item-right docblock-short"><p>Functionality for performing common operations with assets.</p>
</div></div><div class="item-row"><div class="item-left Module-item"><a class="mod" href="asset_id/index.html">asset_id</a></div><div class="item-right docblock-short"><p>The <code>AssetId</code> type used for interacting with an asset on the fuel network.</p>
</div></div><div class="item-row"><div class="item-left Module-item"><a class="mod" href="auth/index.html">auth</a></div><div class="item-right docblock-short"><p>Functionality for determining who is calling a contract.</p>
</div></div><div class="item-row"><div class="item-left Module-item"><a class="mod" href="b512/index.html">b512</a></div><div class="item-right docblock-short"><p>The <code>B512</code> type supports the usage of 64-byte values in Sway which are needed when working with public keys and signatures.</p></div></div><div class="item-row"><div class="item-left Module-item"><a class="mod" href="block/index.html">block</a></div><div class="item-right docblock-short"><p>Functionality for accessing block-related data.</p>
</div></div><div class="item-row"><div class="item-left Module-item"><a class="mod" href="bytes/index.html">bytes</a></div><div class="item-right docblock-short"><p>The <code>Bytes</code> type is used when a collection of tightly-packed arbitrary bytes is needed.</p></div></div><div class="item-row"><div class="item-left Module-item"><a class="mod" href="call_frames/index.html">call_frames</a></div><div class="item-right docblock-short"><p>Helper functions for accessing data from call frames.<br /></div></div><div class="item-row"><div class="item-left Module-item"><a class="mod" href="clone/index.html">clone</a></div><div class="item-right docblock-short"><p>The clone trait, for explicit duplication.</p>
</div></div><div class="item-row"><div class="item-left Module-item"><a class="mod" href="codec/index.html">codec</a></div></div><div class="item-row"><div class="item-left Module-item"><a class="mod" href="constants/index.html">constants</a></div><div class="item-right docblock-short"><p>Definitions for constant values in Sway.</p>
</div></div><div class="item-row"><div class="item-left Module-item"><a class="mod" href="context/index.html">context</a></div><div class="item-right docblock-short"><p>Functionality for accessing context-specific information about the current contract or message.</p></div></div><div class="item-row"><div class="item-left Module-item"><a class="mod" href="contract_id/index.html">contract_id</a></div><div class="item-right docblock-short"><p>The <code>ContractId</code> type used for interacting with contracts on the fuel network.</p>
</div></div><div class="item-row"><div class="item-left Module-item"><a class="mod" href="convert/index.html">convert</a></div><div class="item-right docblock-short"><p>Traits for conversions between types.</p>
</div></div><div class="item-row"><div class="item-left Module-item"><a class="mod" href="crypto/index.html">crypto</a></div></div><div class="item-row"><div class="item-left Module-item"><a class="mod" href="debug/index.html">debug</a></div></div><div class="item-row"><div class="item-left Module-item"><a class="mod" href="ecr/index.html">ecr</a></div><div class="item-right docblock-short"><p>Helper functions to verify signatures.</p>
</div></div><div class="item-row"><div class="item-left Module-item"><a class="mod" href="error_signals/index.html">error_signals</a></div><div class="item-right docblock-short"><p>Values which signify special types of errors when passed to <code>std::revert::revert</code>.</p></div></div><div class="item-row"><div class="item-left Module-item"><a class="mod" href="execution/index.html">execution</a></div><div class="item-right docblock-short"><p>Helper functions to load and run external contract code.</p>
</div></div><div class="item-row"><div class="item-left Module-item"><a class="mod" href="external/index.html">external</a></div><div class="item-right docblock-short"><p>Functions to work with external contracts.</p>
</div></div><div class="item-row"><div class="item-left Module-item"><a class="mod" href="flags/index.html">flags</a></div><div class="item-right docblock-short"><p>Functionality for setting and unsetting FuelVM flags to modify behavior related to the <code>$err</code> and <code>$of</code> registers.</p></div></div><div class="item-row"><div class="item-left Module-item"><a class="mod" href="hash/index.html">hash</a></div><div class="item-right docblock-short"><p>Utility functions for cryptographic hashing.</p>
</div></div><div class="item-row"><div class="item-left Module-item"><a class="mod" href="identity/index.html">identity</a></div><div class="item-right docblock-short"><p>A wrapper type with two variants, <code>Address</code> and <code>ContractId</code>.<br /></div></div><div class="item-row"><div class="item-left Module-item"><a class="mod" href="inputs/index.html">inputs</a></div><div class="item-right docblock-short"><p>Getters for fields on transaction inputs.<br /></div></div><div class="item-row"><div class="item-left Module-item"><a class="mod" href="intrinsics/index.html">intrinsics</a></div><div class="item-right docblock-short"><p>Exposes compiler intrinsics as stdlib wrapper functions.</p>
</div></div><div class="item-row"><div class="item-left Module-item"><a class="mod" href="iterator/index.html">iterator</a></div><div class="item-right docblock-short"><p>The iterator trait to iterate over elements.</p>
</div></div><div class="item-row"><div class="item-left Module-item"><a class="mod" href="logging/index.html">logging</a></div><div class="item-right docblock-short"><p>Allows logging of arbitrary stack types, emitted as either <code>Log</code> or <code>Logd</code> receipts.</p></div></div><div class="item-row"><div class="item-left Module-item"><a class="mod" href="low_level_call/index.html">low_level_call</a></div><div class="item-right docblock-short"><p>Utilities to help with low level calls.</p>
</div></div><div class="item-row"><div class="item-left Module-item"><a class="mod" href="marker/index.html">marker</a></div><div class="item-right docblock-short"><p>Marker traits that represent certain properties of types.</p>
</div></div><div class="item-row"><div class="item-left Module-item"><a class="mod" href="math/index.html">math</a></div><div class="item-right docblock-short"><p>Utilities for common math operations.</p>
</div></div><div class="item-row"><div class="item-left Module-item"><a class="mod" href="message/index.html">message</a></div><div class="item-right docblock-short"><p>Helper functions to sign and send messages.</p>
</div></div><div class="item-row"><div class="item-left Module-item"><a class="mod" href="ops/index.html">ops</a></div></div><div class="item-row"><div class="item-left Module-item"><a class="mod" href="option/index.html">option</a></div><div class="item-right docblock-short"><p>A type for optional values.</p>
</div></div><div class="item-row"><div class="item-left Module-item"><a class="mod" href="outputs/index.html">outputs</a></div><div class="item-right docblock-short"><p>Getters for fields on transaction outputs.<br /></div></div><div class="item-row"><div class="item-left Module-item"><a class="mod" href="raw_slice/index.html">raw_slice</a></div></div><div class="item-row"><div class="item-left Module-item"><a class="mod" href="registers/index.html">registers</a></div><div class="item-right docblock-short"><p>Functions to expose 14 of the reserved FuelVM registers for ease of use.<br /></div></div><div class="item-row"><div class="item-left Module-item"><a class="mod" href="result/index.html">result</a></div><div class="item-right docblock-short"><p>Error handling with the <code>Result</code> type.</p>
</div></div><div class="item-row"><div class="item-left Module-item"><a class="mod" href="revert/index.html">revert</a></div><div class="item-right docblock-short"><p>Functions to panic or revert with a given error code.</p>
</div></div><div class="item-row"><div class="item-left Module-item"><a class="mod" href="storage/index.html">storage</a></div><div class="item-right docblock-short"><p>Contract storage utilities.</p>
</div></div><div class="item-row"><div class="item-left Module-item"><a class="mod" href="str/index.html">str</a></div></div><div class="item-row"><div class="item-left Module-item"><a class="mod" href="string/index.html">string</a></div><div class="item-right docblock-short"><p>A UTF-8 encoded growable string.</p>
</div></div><div class="item-row"><div class="item-left Module-item"><a class="mod" href="time/index.html">time</a></div></div><div class="item-row"><div class="item-left Module-item"><a class="mod" href="tx/index.html">tx</a></div><div class="item-right docblock-short"><p>Transaction field getters.</p>
</div></div><div class="item-row"><div class="item-left Module-item"><a class="mod" href="u128/index.html">u128</a></div><div class="item-right docblock-short"><p>A 128-bit unsigned integer type.</p>
</div></div><div class="item-row"><div class="item-left Module-item"><a class="mod" href="vec/index.html">vec</a></div><div class="item-right docblock-short"><p>A vector type for dynamically sized arrays outside of storage.</p>
</div></div><div class="item-row"><div class="item-left Module-item"><a class="mod" href="vm/index.html">vm</a></div><div class="item-right docblock-short"><p>VM-specific utilities.</p>
</div></div></div><h2 id="primitives">Primitives</h2><div class="item-table"><div class="item-row"><div class="item-left Primitive-item"><a class="primitive" href="primitive.b256.html">b256</a></div><div class="item-right docblock-short">256 bits (32 bytes), i.e. a hash</div></div><div class="item-row"><div class="item-left Primitive-item"><a class="primitive" href="primitive.bool.html">bool</a></div><div class="item-right docblock-short">Boolean true or false</div></div><div class="item-row"><div class="item-left Primitive-item"><a class="primitive" href="primitive.str.html">str</a></div><div class="item-right docblock-short">string slice</div></div><div class="item-row"><div class="item-left Primitive-item"><a class="primitive" href="primitive.str[0].html">str[0]</a></div><div class="item-right docblock-short">fixed-length string</div></div><div class="item-row"><div class="item-left Primitive-item"><a class="primitive" href="primitive.str[10].html">str[10]</a></div><div class="item-right docblock-short">fixed-length string</div></div><div class="item-row"><div class="item-left Primitive-item"><a class="primitive" href="primitive.str[11].html">str[11]</a></div><div class="item-right docblock-short">fixed-length string</div></div><div class="item-row"><div class="item-left Primitive-item"><a class="primitive" href="primitive.str[12].html">str[12]</a></div><div class="item-right docblock-short">fixed-length string</div></div><div class="item-row"><div class="item-left Primitive-item"><a class="primitive" href="primitive.str[13].html">str[13]</a></div><div class="item-right docblock-short">fixed-length string</div></div><div class="item-row"><div class="item-left Primitive-item"><a class="primitive" href="primitive.str[14].html">str[14]</a></div><div class="item-right docblock-short">fixed-length string</div></div><div class="item-row"><div class="item-left Primitive-item"><a class="primitive" href="primitive.str[15].html">str[15]</a></div><div class="item-right docblock-short">fixed-length string</div></div><div class="item-row"><div class="item-left Primitive-item"><a class="primitive" href="primitive.str[16].html">str[16]</a></div><div class="item-right docblock-short">fixed-length string</div></div><div class="item-row"><div class="item-left Primitive-item"><a class="primitive" href="primitive.str[17].html">str[17]</a></div><div class="item-right docblock-short">fixed-length string</div></div><div class="item-row"><div class="item-left Primitive-item"><a class="primitive" href="primitive.str[18].html">str[18]</a></div><div class="item-right docblock-short">fixed-length string</div></div><div class="item-row"><div class="item-left Primitive-item"><a class="primitive" href="primitive.str[19].html">str[19]</a></div><div class="item-right docblock-short">fixed-length string</div></div><div class="item-row"><div class="item-left Primitive-item"><a class="primitive" href="primitive.str[1].html">str[1]</a></div><div class="item-right docblock-short">fixed-length string</div></div><div class="item-row"><div class="item-left Primitive-item"><a class="primitive" href="primitive.str[20].html">str[20]</a></div><div class="item-right docblock-short">fixed-length string</div></div><div class="item-row"><div class="item-left Primitive-item"><a class="primitive" href="primitive.str[21].html">str[21]</a></div><div class="item-right docblock-short">fixed-length string</div></div><div class="item-row"><div class="item-left Primitive-item"><a class="primitive" href="primitive.str[22].html">str[22]</a></div><div class="item-right docblock-short">fixed-length string</div></div><div class="item-row"><div class="item-left Primitive-item"><a class="primitive" href="primitive.str[23].html">str[23]</a></div><div class="item-right docblock-short">fixed-length string</div></div><div class="item-row"><div class="item-left Primitive-item"><a class="primitive" href="primitive.str[24].html">str[24]</a></div><div class="item-right docblock-short">fixed-length string</div></div><div class="item-row"><div class="item-left Primitive-item"><a class="primitive" href="primitive.str[25].html">str[25]</a></div><div class="item-right docblock-short">fixed-length string</div></div><div class="item-row"><div class="item-left Primitive-item"><a class="primitive" href="primitive.str[26].html">str[26]</a></div><div class="item-right docblock-short">fixed-length string</div></div><div class="item-row"><div class="item-left Primitive-item"><a class="primitive" href="primitive.str[27].html">str[27]</a></div><div class="item-right docblock-short">fixed-length string</div></div><div class="item-row"><div class="item-left Primitive-item"><a class="primitive" href="primitive.str[28].html">str[28]</a></div><div class="item-right docblock-short">fixed-length string</div></div><div class="item-row"><div class="item-left Primitive-item"><a class="primitive" href="primitive.str[29].html">str[29]</a></div><div class="item-right docblock-short">fixed-length string</div></div><div class="item-row"><div class="item-left Primitive-item"><a class="primitive" href="primitive.str[2].html">str[2]</a></div><div class="item-right docblock-short">fixed-length string</div></div><div class="item-row"><div class="item-left Primitive-item"><a class="primitive" href="primitive.str[30].html">str[30]</a></div><div class="item-right docblock-short">fixed-length string</div></div><div class="item-row"><div class="item-left Primitive-item"><a class="primitive" href="primitive.str[31].html">str[31]</a></div><div class="item-right docblock-short">fixed-length string</div></div><div class="item-row"><div class="item-left Primitive-item"><a class="primitive" href="primitive.str[32].html">str[32]</a></div><div class="item-right docblock-short">fixed-length string</div></div><div class="item-row"><div class="item-left Primitive-item"><a class="primitive" href="primitive.str[33].html">str[33]</a></div><div class="item-right docblock-short">fixed-length string</div></div><div class="item-row"><div class="item-left Primitive-item"><a class="primitive" href="primitive.str[34].html">str[34]</a></div><div class="item-right docblock-short">fixed-length string</div></div><div class="item-row"><div class="item-left Primitive-item"><a class="primitive" href="primitive.str[35].html">str[35]</a></div><div class="item-right docblock-short">fixed-length string</div></div><div class="item-row"><div class="item-left Primitive-item"><a class="primitive" href="primitive.str[36].html">str[36]</a></div><div class="item-right docblock-short">fixed-length string</div></div><div class="item-row"><div class="item-left Primitive-item"><a class="primitive" href="primitive.str[37].html">str[37]</a></div><div class="item-right docblock-short">fixed-length string</div></div><div class="item-row"><div class="item-left Primitive-item"><a class="primitive" href="primitive.str[38].html">str[38]</a></div><div class="item-right docblock-short">fixed-length string</div></div><div class="item-row"><div class="item-left Primitive-item"><a class="primitive" href="primitive.str[39].html">str[39]</a></div><div class="item-right docblock-short">fixed-length string</div></div><div class="item-row"><div class="item-left Primitive-item"><a class="primitive" href="primitive.str[3].html">str[3]</a></div><div class="item-right docblock-short">fixed-length string</div></div><div class="item-row"><div class="item-left Primitive-item"><a class="primitive" href="primitive.str[40].html">str[40]</a></div><div class="item-right docblock-short">fixed-length string</div></div><div class="item-row"><div class="item-left Primitive-item"><a class="primitive" href="primitive.str[41].html">str[41]</a></div><div class="item-right docblock-short">fixed-length string</div></div><div class="item-row"><div class="item-left Primitive-item"><a class="primitive" href="primitive.str[42].html">str[42]</a></div><div class="item-right docblock-short">fixed-length string</div></div><div class="item-row"><div class="item-left Primitive-item"><a class="primitive" href="primitive.str[43].html">str[43]</a></div><div class="item-right docblock-short">fixed-length string</div></div><div class="item-row"><div class="item-left Primitive-item"><a class="primitive" href="primitive.str[44].html">str[44]</a></div><div class="item-right docblock-short">fixed-length string</div></div><div class="item-row"><div class="item-left Primitive-item"><a class="primitive" href="primitive.str[45].html">str[45]</a></div><div class="item-right docblock-short">fixed-length string</div></div><div class="item-row"><div class="item-left Primitive-item"><a class="primitive" href="primitive.str[46].html">str[46]</a></div><div class="item-right docblock-short">fixed-length string</div></div><div class="item-row"><div class="item-left Primitive-item"><a class="primitive" href="primitive.str[47].html">str[47]</a></div><div class="item-right docblock-short">fixed-length string</div></div><div class="item-row"><div class="item-left Primitive-item"><a class="primitive" href="primitive.str[48].html">str[48]</a></div><div class="item-right docblock-short">fixed-length string</div></div><div class="item-row"><div class="item-left Primitive-item"><a class="primitive" href="primitive.str[49].html">str[49]</a></div><div class="item-right docblock-short">fixed-length string</div></div><div class="item-row"><div class="item-left Primitive-item"><a class="primitive" href="primitive.str[4].html">str[4]</a></div><div class="item-right docblock-short">fixed-length string</div></div><div class="item-row"><div class="item-left Primitive-item"><a class="primitive" href="primitive.str[50].html">str[50]</a></div><div class="item-right docblock-short">fixed-length string</div></div><div class="item-row"><div class="item-left Primitive-item"><a class="primitive" href="primitive.str[51].html">str[51]</a></div><div class="item-right docblock-short">fixed-length string</div></div><div class="item-row"><div class="item-left Primitive-item"><a class="primitive" href="primitive.str[52].html">str[52]</a></div><div class="item-right docblock-short">fixed-length string</div></div><div class="item-row"><div class="item-left Primitive-item"><a class="primitive" href="primitive.str[53].html">str[53]</a></div><div class="item-right docblock-short">fixed-length string</div></div><div class="item-row"><div class="item-left Primitive-item"><a class="primitive" href="primitive.str[54].html">str[54]</a></div><div class="item-right docblock-short">fixed-length string</div></div><div class="item-row"><div class="item-left Primitive-item"><a class="primitive" href="primitive.str[55].html">str[55]</a></div><div class="item-right docblock-short">fixed-length string</div></div><div class="item-row"><div class="item-left Primitive-item"><a class="primitive" href="primitive.str[56].html">str[56]</a></div><div class="item-right docblock-short">fixed-length string</div></div><div class="item-row"><div class="item-left Primitive-item"><a class="primitive" href="primitive.str[57].html">str[57]</a></div><div class="item-right docblock-short">fixed-length string</div></div><div class="item-row"><div class="item-left Primitive-item"><a class="primitive" href="primitive.str[58].html">str[58]</a></div><div class="item-right docblock-short">fixed-length string</div></div><div class="item-row"><div class="item-left Primitive-item"><a class="primitive" href="primitive.str[59].html">str[59]</a></div><div class="item-right docblock-short">fixed-length string</div></div><div class="item-row"><div class="item-left Primitive-item"><a class="primitive" href="primitive.str[5].html">str[5]</a></div><div class="item-right docblock-short">fixed-length string</div></div><div class="item-row"><div class="item-left Primitive-item"><a class="primitive" href="primitive.str[60].html">str[60]</a></div><div class="item-right docblock-short">fixed-length string</div></div><div class="item-row"><div class="item-left Primitive-item"><a class="primitive" href="primitive.str[61].html">str[61]</a></div><div class="item-right docblock-short">fixed-length string</div></div><div class="item-row"><div class="item-left Primitive-item"><a class="primitive" href="primitive.str[62].html">str[62]</a></div><div class="item-right docblock-short">fixed-length string</div></div><div class="item-row"><div class="item-left Primitive-item"><a class="primitive" href="primitive.str[63].html">str[63]</a></div><div class="item-right docblock-short">fixed-length string</div></div><div class="item-row"><div class="item-left Primitive-item"><a class="primitive" href="primitive.str[64].html">str[64]</a></div><div class="item-right docblock-short">fixed-length string</div></div><div class="item-row"><div class="item-left Primitive-item"><a class="primitive" href="primitive.str[6].html">str[6]</a></div><div class="item-right docblock-short">fixed-length string</div></div><div class="item-row"><div class="item-left Primitive-item"><a class="primitive" href="primitive.str[7].html">str[7]</a></div><div class="item-right docblock-short">fixed-length string</div></div><div class="item-row"><div class="item-left Primitive-item"><a class="primitive" href="primitive.str[8].html">str[8]</a></div><div class="item-right docblock-short">fixed-length string</div></div><div class="item-row"><div class="item-left Primitive-item"><a class="primitive" href="primitive.str[9].html">str[9]</a></div><div class="item-right docblock-short">fixed-length string</div></div><div class="item-row"><div class="item-left Primitive-item"><a class="primitive" href="primitive.u16.html">u16</a></div><div class="item-right docblock-short">16-bit unsigned integer</div></div><div class="item-row"><div class="item-left Primitive-item"><a class="primitive" href="primitive.u256.html">u256</a></div><div class="item-right docblock-short">256-bit unsigned integer</div></div><div class="item-row"><div class="item-left Primitive-item"><a class="primitive" href="primitive.u32.html">u32</a></div><div class="item-right docblock-short">32-bit unsigned integer</div></div><div class="item-row"><div class="item-left Primitive-item"><a class="primitive" href="primitive.u64.html">u64</a></div><div class="item-right docblock-short">64-bit unsigned integer</div></div><div class="item-row"><div class="item-left Primitive-item"><a class="primitive" href="primitive.u8.html">u8</a></div><div class="item-right docblock-short">8-bit unsigned integer</div></div></div></section><section id="search" class="search-results"></section></div></main><script src="../static.files/highlight.js"></script><script>hljs.highlightAll();</script></body></html>