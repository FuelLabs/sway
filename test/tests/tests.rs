use libtest_mimic::{Arguments, Trial};
use normalize_path::NormalizePath;
use regex::Regex;
use std::{path::PathBuf, str::FromStr, sync::Once};

static FORC_COMPILATION: Once = Once::new();
static FORC_DOC_COMPILATION: Once = Once::new();

fn compile_forc() {
    let args = vec!["b", "--release", "-p", "forc"];
    let o = std::process::Command::new("cargo")
        .args(args)
        .output()
        .unwrap();
    assert!(o.status.success());
}

fn compile_forc_doc() {
    let args = vec!["b", "--release", "-p", "forc-doc"];
    let o = std::process::Command::new("cargo")
        .args(args)
        .output()
        .unwrap();
    assert!(o.status.success());
}

pub fn main() {
    let repo_root: PathBuf =
        PathBuf::from_str(&std::env::var("CARGO_MANIFEST_DIR").unwrap()).unwrap();
    let repo_root = repo_root.parent().unwrap().to_path_buf();

    let mut args = Arguments::from_args();
    args.nocapture = true;

    let tests = discover_test()
        .into_iter()
        .map(|dir| {
            let test_programs_dir = "src/e2e_vm_tests/test_programs/";
            let name = dir
                .to_str()
                .unwrap()
                .to_string()
                .replace(test_programs_dir, "");

            let repo_root = repo_root.clone();
            Trial::test(name, move || {
                let snapshot_toml =
                    std::fs::read_to_string(format!("{}/snapshot.toml", dir.display()))?;
                let snapshot_toml = toml::from_str::<toml::Value>(&snapshot_toml)?;
                let cmds = if let Some(cmds) = snapshot_toml.get("cmds") {
                    cmds.as_array()
                        .unwrap()
                        .iter()
                        .map(|v| v.as_str().unwrap())
                        .collect::<Vec<_>>()
                } else {
                    vec!["forc build --path {root}"]
                };

                let root = format!("test/{}", dir.display());

                use std::fmt::Write;
                let mut snapshot = String::new();

                for cmd in cmds {
                    let cmd = cmd.replace("{root}", &root);

                    let _ = writeln!(&mut snapshot, "> {}", cmd);

                    // known commands
                    let cmd = if let Some(cmd) = cmd.strip_prefix("forc doc ") {
                        FORC_DOC_COMPILATION.call_once(|| {
                            compile_forc_doc();
                        });
                        format!("target/release/forc-doc {cmd} 1>&2")
                    } else if let Some(cmd) = cmd.strip_prefix("forc ") {
                        FORC_COMPILATION.call_once(|| {
                            compile_forc();
                        });
                        format!("target/release/forc {cmd} 1>&2")
                    } else {
                        panic!("Not supported. Possible commands: forc")
                    };

                    let o = duct::cmd!("bash", "-c", cmd.clone())
                        .dir(repo_root.clone())
                        .stderr_to_stdout()
                        .stdout_capture()
                        .env("COLUMNS", "10")
                        .unchecked()
                        .start()
                        .unwrap();
                    let o = o.wait().unwrap();

                    let _ = writeln!(
                        &mut snapshot,
                        "{}",
                        clean_output(&format!(
                            "exit status: {}\noutput:\n{}",
                            o.status.code().unwrap(),
                            std::str::from_utf8(&o.stdout).unwrap(),
                        ))
                    );
                }

                fn stdout(root: &str, snapshot: &str) {
                    let root = PathBuf::from_str(root).unwrap();
                    let root = root.normalize();

                    let mut insta = insta::Settings::new();
                    insta.set_snapshot_path(root);
                    insta.set_prepend_module_to_snapshot(false);
                    insta.set_omit_expression(true);
                    let scope = insta.bind_to_scope();
                    insta::assert_snapshot!("stdout", snapshot);
                    drop(scope);
                }
                stdout(&format!("{}/{root}", repo_root.display()), &snapshot);

                Ok(())
            })
        })
        .collect();
    libtest_mimic::run(&args, tests).exit();
}

pub fn discover_test() -> Vec<PathBuf> {
    use glob::glob;

    let mut entries = vec![];

    for entry in glob("**/snapshot.toml")
        .expect("Failed to read glob pattern")
        .flatten()
    {
        entries.push(entry.parent().unwrap().to_owned())
    }

    entries
}

fn clean_output(output: &str) -> String {
    #[derive(Default)]
    struct RawText(String);

    impl vte::Perform for RawText {
        fn print(&mut self, c: char) {
            self.0.push(c);
        }

        fn execute(&mut self, _: u8) {}

        fn hook(&mut self, _: &vte::Params, _: &[u8], _: bool, _: char) {}

        fn put(&mut self, b: u8) {
            self.0.push(b as char);
        }

        fn unhook(&mut self) {}

        fn osc_dispatch(&mut self, _: &[&[u8]], _: bool) {}

        fn csi_dispatch(&mut self, _: &vte::Params, _: &[u8], _: bool, _: char) {}

        fn esc_dispatch(&mut self, _: &[u8], _: bool, _: u8) {}
    }

    let mut raw = String::new();
    for line in output.lines() {
        let mut performer = RawText::default();
        let mut p = vte::Parser::new();
        for b in line.as_bytes() {
            p.advance(&mut performer, *b);
        }
        raw.push_str(&performer.0);
        raw.push('\n');
    }

    // Remove absolute paths from snapshot tests
    let manifest_dir = env!("CARGO_MANIFEST_DIR");
    let manifest_dir: PathBuf = PathBuf::from(manifest_dir);
    let parent = manifest_dir.parent().unwrap();
    let result = raw.replace(&format!("{}/", parent.display()), "");

    // Remove compilation time
    let r = Regex::new("(Finished release \\[.*?\\] target\\(s\\) \\[.*?\\] in )(.*?s)").unwrap();
    let result = r.replace(&result, "$1???");

    result.to_string()
}
