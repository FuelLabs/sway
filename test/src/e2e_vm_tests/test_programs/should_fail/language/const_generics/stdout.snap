---
source: test/tests/tests.rs
---
> forc build --path test/src/e2e_vm_tests/test_programs/should_fail/language/const_generics
exit status: 1
output:
    Building test/src/e2e_vm_tests/test_programs/should_fail/language/const_generics
   Compiling library core (sway-lib-core)
   Compiling library std (sway-lib-std)
   Compiling script const_generics (test/src/e2e_vm_tests/test_programs/should_fail/language/const_generics)
error
 --> test/src/e2e_vm_tests/test_programs/should_fail/language/const_generics/src/main.sw:7:15
  |
5 | }
6 | 
7 | impl<T, const N: u64> A for [T; N] {
  |               ^ This needs "const_generics" to be enabled, but it is currently disabled. For more details go to https://github.com/FuelLabs/sway/issues/6860.
8 |     #[allow(dead_code)]
9 |     fn f(self) -> u64 {
  |
____

  Aborting due to 1 error.
error: Failed to compile const_generics

> forc build --path test/src/e2e_vm_tests/test_programs/should_fail/language/const_generics --experimental const_generics
exit status: 1
output:
    Building test/src/e2e_vm_tests/test_programs/should_fail/language/const_generics
   Compiling library core (sway-lib-core)
   Compiling library std (sway-lib-std)
   Compiling script const_generics (test/src/e2e_vm_tests/test_programs/should_fail/language/const_generics)
error
 --> test/src/e2e_vm_tests/test_programs/should_fail/language/const_generics/src/main.sw:7:15
  |
5 | }
6 | 
7 | impl<T, const N: u64> A for [T; N] {
  |               ^ "const generics" are not supported here.
8 |     #[allow(dead_code)]
9 |     fn f(self) -> u64 {
  |
____

  Aborting due to 1 error.
error: Failed to compile const_generics
